src/coreclr/System.Private.CoreLib/src/System/Runtime/CompilerServices/RuntimeHelpers.CoreCLR.cs

```diff
@@ -750,11 +842,39 @@ public bool IsTypeDesc
             return (MethodTable*)m_asTAddr;
         }
 
+        /// <summary>
+        /// Gets the <see cref="TypeDesc"/> pointer wrapped by the current instance.
+        /// </summary>
+        /// <remarks>This is only safe to call if <see cref="IsTypeDesc"/> returned <see langword="true"/>.</remarks>
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public TypeDesc* AsTypeDesc()
+        {
+            Debug.Assert(IsTypeDesc);
+
+            return (TypeDesc*)((nint)m_asTAddr - 2);
+        }
+
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static TypeHandle TypeHandleOf<T>()
         {
             return new TypeHandle((void*)RuntimeTypeHandle.ToIntPtr(typeof(T).TypeHandle));
         }
+
+        public static bool AreSameType(TypeHandle left, TypeHandle right) => left.m_asTAddr == right.m_asTAddr;
+
+        public CorElementType GetVerifierCorElementType() => IsTypeDesc
+            ? AsTypeDesc()->GetInternalCorElementType()
```

Comment:
Since `MethodTable::GetVerifierCorElementType` is FCall, there is not much value in duplicating TypeDesc::GetVerifierCorElementType() in C#. I would make FCall for the whole `TypeHandle::GetVerifierCorElementType` instead.

Reply:
GetVerifierCorElementType() seems unused on managed side. Maybe just delete this FCall and keep implementation on native side (for jitinterface, arraynative etc.)?

Reply:
It's used for deciding primitive widen. Some other usages were refactored away, but `GetCorElementType` is using it.

Reply:
Well, the remaining CorElementType usages are about primitive widen and fast path of `int[]` <-> `uint[]` conversion. It's only used for primitive-like types. In these cases, `GetVerifierCorElementType` has unnecessary overhead, and `Enum` has defined a custom FCall.

It think the `GetVerifierCorElementType` FCall should be removed and share a new FCall for primitive-like types with Enum. Is it correct that a `TypeHandle` can only be primitive when it's `MethodTable` and the primitive flag is set? Are (U)IntPtr always represented as `MethodTable` with ELEMENT_TYPE_I/U?

Reply:
Well all of the usages of `GetCorElementTypeOfElementType` are in fact about primitive types. It can be refactored later.